ports/geckolib/gecko_bindings/sugar/refptr.rs

```diff
@@ -0,0 +1,91 @@
+use std::marker::PhantomData;
+use std::mem::{forget, transmute};
+use std::ptr::{read, null};
+use std::sync::Arc;
+
+/// Helper trait for conversions between FFI Strong/Borrowed types and Arcs
+pub unsafe trait RefCounted where Self: Sized {
+    /// Gecko's name for the type
+    /// This is equivalent to ArcBox<Self>
+    type FFIType: Sized;
+    fn with<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output
```

Comment:
Can we really mark this as safe? It starts with a debug_assert! which could fail, and invokes UB if that debug_assert! fails. I would want to either assert! this, or mark it as unsafe.

I would also prefer naming the unsafe version `with_unchecked`, and the safe one just `with`.

I know that in gecko we dereference pointers without checking if they are null all of the time, but that's a bad practice, and we should avoid doing it in rust code unless we have profiles showing that we need that extra performance.